### DISCUSSION TITLE
Communication Between Prisma Layers

### DISCUSSION THREAD

**Discussion by queirozlc** [prisma/prisma#22623] [Q&A] [answered]
### Question

Hello guys 👋

Recently i started to digging into prisma source code in order to understand the architecture and what exactly happens under the hood. What i want to know is how the javascript layer of prisma communicates with the engine layer? Does the engineType have some influence over this communication ? What's the difference between choose a library engine or a binary engine type? 

I'd love to contribute with prisma project but i feel that could be awesome if i was able to understand the entire flow of this project.

Sorry if my question is not according to the expected  😅

**Comment by janpio** [comment, marked as answer]
The communication depends on the `engineType`. Without specifying that property or with the default `library` we use a Node-API based library, with `binary` a executable binary start starts a HTTP server. Both API get a Json payload from the Prisma Client library with the Client query to execute, and also return the data as Json after the Query engine did its job.

**Reply by queirozlc** [reply]
Thanks for the answer mate! Other question i've had:

On library mode, the binaries of all engine source code are compiled as a Node-API ? Even the rust code of `psl` for example? Or just the binaries of `query-engine` or `migration-engine`?

**Reply by janpio** [reply]
The `engineType` only influences Prisma Client, which uses the `query-engine`.
The `schema-engine` is always compiled as a executable that is used by Prisma CLI.
And then we also have a few Wasm modules that are used by the CLI and the VS Code extension.

`psl` on the other hand is just an internal crate, that is used by the `query-engine` and `schema-engine` under the hood. We have quite a few of those.

(`migration-engine` and `introspection` are the old names of two separate engines that were later then combined to form the new `schema-engine` which does what both of these did before)

**Reply by queirozlc** [reply]
Enlightening, hope i can  do some contribution someday for this library, thanks mate 🫡

**Reply by rhyek** [reply]
@janpio do you know if there's any performance considerations when choosing between `binary` and `library`? Is one meant to be more performant than the other? It sounds like `library` may be more performant from your explanation.

I wonder if there's any benchmarks that could be useful.

**Reply by rhyek** [reply]
I found this at https://www.prisma.io/docs/orm/more/under-the-hood/engines#the-query-engine-at-runtime
> The Node-API library approach is recommended since it reduces the communication overhead between Prisma Client and the query engine.
